### PR TITLE
Alerting stage 2

### PR DESCRIPTION
**Pull request by julmon**
# HTTP API

## Current state

```console
$ curl -s -k --cookie "temboard=2|1:0|10:1523862974|8:temboard|32:MjI6YWRtaW46eGlldDdLTHVtdXg1MFEy|54972bb4107b0cee156b55bbe200a9c507235350c358fce81f06a2072cad36dd" \
https://temboard-ui-centos7:8888/server/192.168.122.21/2345/alerting/json/state

[{"state": "OK", "name": "archive_ready", "key": ""}, {"state": "WARNING", "name": "cpu", "key": "cpu0"}, {"state": "OK", "name": "fs", "key": "/"}, {"state": "OK", "name": "fs", "key": "/boot"}, {"state": "OK", "name": "fs", "key": "/dev"}, {"state": "OK", "name": "fs", "key": "/dev/shm"}, {"state": "OK", "name": "fs", "key": "/run"}, {"state": "OK", "name": "fs", "key": "/run/user/0"}, {"state": "OK", "name": "fs", "key": "/run/user/1000"}, {"state": "OK", "name": "fs", "key": "/sys/fs/cgroup"}, {"state": "OK", "name": "hitratio", "key": "bench"}, {"state": "OK", "name": "hitratio", "key": "postgres"}, {"state": "OK", "name": "hitratio", "key": "temboard"}, {"state": "OK", "name": "hitratio", "key": "template1"}, {"state": "CRITICAL", "name": "loadaverage", "key": ""}, {"state": "OK", "name": "memory", "key": ""}, {"state": "OK", "name": "sessions", "key": ""}, {"state": "OK", "name": "swap", "key": ""}, {"state": "OK", "name": "waiting", "key": "bench"}, {"state": "OK", "name": "waiting", "key": "postgres"}, {"state": "OK", "name": "waiting", "key": "temboard"}, {"state": "OK", "name": "waiting", "key": "template1"}, {"state": "OK", "name": "wal_files", "key": ""}, {"state": "OK", "name": "xacts_rollback", "key": "bench"}, {"state": "OK", "name": "xacts_rollback", "key": "postgres"}, {"state": "OK", "name": "xacts_rollback", "key": "temboard"}, {"state": "OK", "name": "xacts_rollback", "key": "template1"}]
```

## Checks list

```console
$ curl -s -k --cookie "temboard=2|1:0|10:1523862974|8:temboard|32:MjI6YWRtaW46eGlldDdLTHVtdXg1MFEy|54972bb4107b0cee156b55bbe200a9c507235350c358fce81f06a2072cad36dd" \
https://temboard-ui-centos7:8888/server/192.168.122.21/2345/alerting/json/checks                                                            

[{"description": "WAL files ready to be archived", "warning": 10.0, "enabled": true, "name": "archive_ready", "criticial": 20.0}, {"description": "CPU usage", "warning": 50.0, "enabled": true, "name": "cpu", "criticial": 80.0}, {"description": "File systems usage (%)", "warning": 80.0, "enabled": true, "name": "fs", "criticial": 90.0}, {"description": "Cache Hit Ratio", "warning": 90.0, "enabled": true, "name": "hitratio", "criticial": 80.0}, {"description": "Loadaverage", "warning": 0.5, "enabled": true, "name": "loadaverage", "criticial": 1.0}, {"description": "Memory usage", "warning": 50.0, "enabled": true, "name": "memory", "criticial": 80.0}, {"description": "Client sessions", "warning": 80.0, "enabled": true, "name": "sessions", "criticial": 90.0}, {"description": "Swap usage", "warning": 30.0, "enabled": true, "name": "swap", "criticial": 50.0}, {"description": "Waiting sessions", "warning": 5.0, "enabled": true, "name": "waiting", "criticial": 10.0}, {"description": "WAL files", "warning": 50.0, "enabled": true, "name": "wal_files", "criticial": 100.0}, {"description": "Rollbacked transactions", "warning": 10.0, "enabled": true, "name": "xacts_rollback", "criticial": 20.0}]
```
## Update Check (`enabled`, `thresholds` or `description`)

```console
$ curl -D - -k --cookie "temboard=2|1:0|10:1523862974|8:temboard|32:MjI6YWRtaW46eGlldDdLTHVtdXg1MFEy|54972bb4107b0cee156b55bbe200a9c507235350c358fce81f06a2072cad36dd" \
-X POST --data '{"checks":[{"name":"fs","enabled":1,"description":"File systems usage (%)", "warning": 70}]}' \
https://temboard-ui-centos7:8888/server/192.168.122.21/2345/alerting/json/checks

HTTP/1.1 200 OK
Date: Tue, 17 Apr 2018 12:13:04 GMT
Content-Length: 2
Content-Type: application/json
Server: TornadoServer/4.5.3

{}
```

## History

```console
$ curl -s -k --cookie "temboard=2|1:0|10:1523862974|8:temboard|32:MjI6YWRtaW46eGlldDdLTHVtdXg1MFEy|54972bb4107b0cee156b55bbe200a9c507235350c358fce81f06a2072cad36dd" \
https://temboard-ui-centos7:8888/server/192.168.122.21/2345/alerting/data/loadaverage\?start\=2018-04-15T00:00:00Z\&end\=2018-04-17T00:00:00Z

datetime,state,key,value,warning,critical
2018-04-17 01:59:35+02,OK,"",0.03,0.5,1
2018-04-17 01:58:35+02,OK,"",0.09,0.5,1
2018-04-17 01:57:34+02,OK,"",0.04,0.5,1
2018-04-17 01:56:34+02,OK,"",0.11,0.5,1
2018-04-17 01:55:34+02,OK,"",0.31,0.5,1
2018-04-17 01:54:33+02,OK,"",0.03,0.5,1
2018-04-17 01:53:33+02,OK,"",0.09,0.5,1
2018-04-17 01:52:32+02,OK,"",0.02,0.5,1
2018-04-17 01:51:32+02,OK,"",0.05,0.5,1
[...]
```
## History with state changes only

```console
$ curl -s -k --cookie "temboard=2|1:0|10:1523862974|8:temboard|32:MjI6YWRtaW46eGlldDdLTHVtdXg1MFEy|54972bb4107b0cee156b55bbe200a9c507235350c358fce81f06a2072cad36dd" \
https://temboard-ui-centos7:8888/server/192.168.122.21/2345/alerting/data/loadaverage\?changes_only\=1\&start\=2018-04-15T00:00:00Z\&end\=2018-04-17T00:00:00Z

datetime,state,key,value,warning,critical
2018-04-16 23:03:27+02,OK,"",0.21,0.5,1
2018-04-16 23:02:27+02,WARNING,"",0.58,0.5,1
2018-04-16 16:43:55+02,OK,"",0.46,0.5,1
2018-04-16 16:37:52+02,CRITICAL,"",1.26,0.5,1
2018-04-16 14:01:49+02,OK,"",0.32,0.5,1
2018-04-16 14:00:49+02,WARNING,"",0.88,0.5,1
2018-04-16 13:56:46+02,CRITICAL,"",1.22,0.5,1
2018-04-16 10:56:23+02,OK,"",0.24,0.5,1
2018-04-16 10:55:23+02,WARNING,"",0.65,0.5,1
2018-04-15 11:20:00+02,OK,"",0.19,0.5,1
2018-04-15 11:19:00+02,WARNING,"",0.51,0.5,1
2018-04-15 07:10:12+02,OK,"",0.19,0.5,1
2018-04-15 07:09:12+02,WARNING,"",0.51,0.5,1
2018-04-15 02:00:10+02,OK,"",0,0.5,1
```

### CODE REVIEW

**Comment by pgiraud** [comment]
What about a service to get the last 50 state changes? ie. using a limit instead of a date range?

File: share/sql/alerting.sql [hunk 1]
```diff
@@ -2,70 +2,138 @@ SET search_path TO monitoring, public;
...
+  PRIMARY KEY (check_id, key)
+);
+
+CREATE TYPE check_results_record AS (
+  datetime TIMESTAMPTZ,
+  state check_state_type,
+  key VARCHAR(64),
+  value REAL,
+  warning REAL,
+  critical REAL
```
Comment (pgiraud) [inline share/sql/alerting.sql]: I would personnaly make it more obvious that `warning` and `critical` are actually thresholds: `warning_threshold` and `critical_threshold`.

File: share/sql/alerting.sql [hunk 2]
```diff
@@ -2,70 +2,138 @@ SET search_path TO monitoring, public;
...
-	threshold_c REAL,
-	description TEXT
+  check_id SERIAL PRIMARY KEY,
+  host_id INTEGER NOT NULL REFERENCES hosts (host_id),
+  instance_id INTEGER REFERENCES instances (instance_id),
+  enabled BOOLEAN NOT NULL DEFAULT false,
+  name VARCHAR(64) NOT NULL,
+  threshold_w REAL,
+  threshold_c REAL,
+  description TEXT
```
Comment (pgiraud) [inline share/sql/alerting.sql]: If SQL indentation is 2 spaces, let's configure it in `.editorconfig`.

File: temboardui/plugins/monitoring/handlers/alerting.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+            instance_id = get_instance_id(self.db_session, host_id,
+                                          instance.pg_port)
+
+            self.db_session.expunge_all()
+
+            # Arguments
+            start = self.get_argument('start', default=None)
+            end = self.get_argument('end', default=None)
+            key = self.get_argument('key', default=None)
+            # Return 200 with empty list when an error occurs
```
Comment (pgiraud) [inline temboardui/plugins/monitoring/handlers/alerting.py]: I don't understand this comment.

File: temboardui/plugins/monitoring/handlers/alerting.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+                                    for plugin in instance.plugins]:
+                raise TemboardUIError(408, "Plugin not active.")
+
+            # Find host_id & instance_id
+            host_id = get_host_id(self.db_session, instance.hostname)
+            instance_id = get_instance_id(self.db_session, host_id,
+                                          instance.pg_port)
+            query = self.db_session.query(
+                        Check.name, Check.enabled, Check.threshold_w,
+                        Check.threshold_c, Check.description
```
Comment (pgiraud) [inline temboardui/plugins/monitoring/handlers/alerting.py]: `self.db_session.query(Check)` would be enough here

**Comment by julmon** [comment]
@pgiraud we need a time frame for getting states history if we don't want to take the risk of reading the whole content of the history (tables `check_results_current` and `check_results_history`)

File: share/sql/alerting.sql (hunk 1 above)
Comment (julmon) [reply share/sql/alerting.sql]: I'd like to keep composite type attributes concise.

File: temboardui/plugins/monitoring/handlers/alerting.py (hunk 3 above)
Comment (julmon) [reply temboardui/plugins/monitoring/handlers/alerting.py]: If `no_error=1` then we do not return HTTP error if something goes wrong.

File: share/sql/alerting.sql (hunk 2 above)
Comment (julmon) [reply share/sql/alerting.sql]: Done with https://github.com/dalibo/temboard/pull/321/commits/0f74c25a02988aaf6fa7eb33f4eea480741ce234

File: .editorconfig [hunk 5]
```diff
@@ -18,3 +18,6 @@ indent_size = 2
 [Makefile]
 indent_size = 8
 indent_style = tab
+
+[*.sql]
+indent_size = 2
```
Comment (pgiraud) [inline .editorconfig]: let's just add this along with css, html, etc...

File: temboardui/plugins/monitoring/handlers/alerting.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+                       (agent_address, agent_port))
+
+
+class AlertingDataCheckResultsHandler(CsvHandler):
+
+    def get_data_check_results(self, address, port, check_name):
+        try:
+            instance = None
+            role = None
+            no_error = 0
```
Comment (pgiraud) [inline temboardui/plugins/monitoring/handlers/alerting.py]: Is this mandatory? You call `get_argument` with `default` keyword argument below.

Comment (julmon) [reply temboardui/plugins/monitoring/handlers/alerting.py]: Nope.

**Pull request merged**